app/src/main/java/co/copperhead/pdfviewer/PdfViewer.java

```diff
@@ -348,34 +394,68 @@ public void onPageFinished(WebView view, String url) {
             mUri = savedInstanceState.getParcelable(STATE_URI);
             mPage = savedInstanceState.getInt(STATE_PAGE);
             mZoomLevel = savedInstanceState.getInt(STATE_ZOOM_LEVEL);
+            mEncryptedDocumentPassword = savedInstanceState.getString(STATE_ENCRYPTED_DOCUMENT_PASSWORD);
         }
 
         if (mUri != null) {
             loadPdf();
         }
     }
 
-    private void loadPdf() {
-        try {
-            if (mInputStream != null) {
-                mInputStream.close();
+    private JumpToPageFragment getJumpToPageFragment() {
+        if (mJumpToPageFragment == null) {
+            mJumpToPageFragment = JumpToPageFragment.newInstance(mPage, mNumPages);
+        }
+        return mJumpToPageFragment;
+    }
+
+    private PasswordPromptFragment getPasswordPromptFragment() {
+        if (mPasswordPromptFragment == null) {
+            final Fragment fragment = getFragmentManager().findFragmentByTag(PasswordPromptFragment.TAG);
+            if (fragment != null) {
+                mPasswordPromptFragment = (PasswordPromptFragment) fragment;
+            } else {
+                mPasswordPromptFragment = new PasswordPromptFragment();
             }
-            mInputStream = getContentResolver().openInputStream(mUri);
-        } catch (IOException e) {
-            return;
         }
+        return mPasswordPromptFragment;
+    }
+
+    private String getDocumentProperties() {
+        if (mDocumentProperties == null) {
+            mDocumentProperties = getString(R.string.document_properties_retrieval_failed);
+        }
+        return mDocumentProperties;
+    }
+
+    private void loadPdf(Uri data) {
+        mDocumentState = 0;
+        mJumpToPageFragment = null;
+        mEncryptedDocumentPassword = null;
+        mInvalidPassword = false;
+        mUri = data;
+        mPage = 1;
+        loadPdf();
+        invalidateOptionsMenu();
+    }
+    
+    private void loadPdf() {
         mWebView.loadUrl("file:///android_asset/viewer.html");
     }
 
+    public void loadPdfWithPassword(String password) {
+        mWebView.evaluateJavascript(String.format("loadDocument(\"%s\")", password), null);
```

Comment:
This should really be changed to passing it via the channel. It's too hard to escape things properly to pass them as arguments via evaluateJavascript.

Reply:
Escaping like that works fine based on my testing, i did the same thing with the option to rotate PDF's, and the only change needed was to parse the arg string to integer.
But if you insist, i'm going to change it.

Reply:
What if it contains a `"` character though?

Reply:
You're right about that (makes sense duh!), my test sample seemed to have ignored non alpha/digit's, i'll add that to my todo list.